Guard tables in compaction tasks

bhalevy opened the pull request:
Currently, if a compaction function enters the table
or compaction_group async_gate, we can't stop it
on the table/compaction_group stop path as they co_await
their respective async_gate.close().

This series introduces a table_ptr smart pointer to guards
the table object by entering its async_gate, and
it also defers awaiting the gate.close future
till after stopping ongoing compaction so that
closing the gate will prevent starting new compactions
while ongoing compaction can be stopped and finally
awaiting the close() future will wait for them to
unwind and exit the gate after being stopped.

Fixes #16305

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5145/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6530/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5145/)
#### Build Details:
- Duration: 2 hr 28 min
- Builder: spider4.cloudius-systems.com

tchaikov commented on replica/database.hh:
```diff
@@ -1180,6 +1180,33 @@ public:
 
+// Smart table pointer that guards the table object
+// while it's being accessed asynchronously
+class table_ptr {
```
nit, i'd suggest rename it to `table_holder` . and add a `hold()` helper to `table` class. the helper constructors calling into `find_column_family()` are nice but not quite necessary.

bhalevy replied:
renamed in v2

tchaikov commented on compaction/task_manager_module.cc:
```diff
@@ -193,7 +193,8 @@ future<> run_on_table(sstring op, replica::database& db, std::string keyspace, t
     tasks::tmlogger.debug("Starting {} on {}.{}", op, keyspace, ti.name);
     try {
-        co_await func(db.find_column_family(ti.id));
+        auto tptr = replica::table_ptr(db, ti.id);
```
nit, the commit message reads:

> To ensure the table will be dropped while ...

i think it misses a "not".

bhalevy replied:
fixed in v2

tchaikov reviewed (approved):
lgtm modulo some nits.

Deexie commented on replica/table.cc:
```diff
@@ -1073,10 +1073,14 @@ table::stop() {
         co_return;
     }
-    co_await _async_gate.close();
+    // Allow `compaction_manager::stop` to stop ongoing compactions
```
`compaction_group::stop`

bhalevy replied:
fixed in v2

Deexie commented on replica/table.cc:
```diff
@@ -1073,10 +1073,14 @@ table::stop() {
-    co_await _async_gate.close();
+    // Allow `compaction_manager::stop` to stop ongoing compactions
+    // while they may still hold the _async_gate
+    // entered in `compaction_manager::start_compaction`
```
`compaction_manager::start_compaction` enters `compaction_state::gate`, but here `table::_async_gate` is closed

bhalevy replied:
right, I'll revise the comment

bhalevy replied:
simplified the comment in v2

Deexie commented on replica/table.cc:
```diff
@@ -1654,9 +1658,13 @@ future<> compaction_group::stop() noexcept {
-    co_await _async_gate.close();
+    // Allow `compaction_manager::remove` to stop ongoing compactions
+    // while they may still hold the _async_gate
+    // entered in `compaction_manager::start_compaction`
```
ditto

bhalevy replied:
no, here it is the compaction_group::_async_gate.
The previous mention was copy-pasted

Deexie replied:
Yes, but it's still a different gate than the one entered in `compaction_manager::start_compaction`

bhalevy replied:
You're right.
Dropped this hunk changing to compacton_group::stop in v2

bhalevy commented:
In v2 ([7843025](https://github.com/scylladb/scylladb/commit/7843025a53fc808120c4f28e578a59817214ce04)):
- table: stop: allow compactions to be stopped while closing async_gate:
  - kept only change in table::stop
  - simplified comment
- table: add table_holder and hold method
  - table_ptr renamed to table_holder
  - dropped db related constructors
  - find_column_family is done by caller (`run_on_table`)
- compaction: run_on_table: hold table
  - adapted to previous patch

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5179/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6572/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5179/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider4.cloudius-systems.com

Deexie reviewed (approved):
LGTM
